designs/inline-io.md

```diff
@@ -0,0 +1,228 @@
+# Inline Operation Inputs / Outputs
+
+This document describes a way to write input and output shapes as an inline
+part of an operation’s definition.
+
+## Motivation
+
+Operation input and output shapes are always structures, almost always have
+boilerplate names, and critically are almost never re-used. In some cases, they
+may not even have types generated for them. Because of those properties, the
+need to fully define them separately from an operation can feel like needless
+boilerplate. Additionally, separating them makes reading an operation at a high
+level more difficult since you have to jump around to get the information you
+need.
+
+## Proposal
+
+Operations will allow inlining input and output definitions, indicated by a
+walrus operator (`:=`). Inlined structures will deviate from normal
+structure definitions in two respects. Firstly, the structure keyword will be
+omitted. Additionally, providing a name for the inlined shape will be optional.
+
+### Walrus Operator
+
+Rather than using the same standalone colon (`:`) that is used in other
+cases, a walrus operator will be used to indicate an inline definition. The
+reason for this is to visually distinguish it at the outset, as well as to make
+tooling a bit cleaner to write.
+
+This usage of the operator is analogous to how some programming languages use
+it. For instance, Python uses it to assign a name to the result of an
+expression in places where initializing a variable was previously forbidden.
+In Go, it’s used to initialize and assign a variable in one step.
+
+### Omitting the structure Keyword
+
+When used at the top level of a Smithy IDL file, the type keywords are
+necessary to indicate what type of shape you’re making. Since operation
+inputs and outputs may only be structures, this isn’t necessary.
+
+### Default Names
```

Comment:
I think we should remove the ability to provide an explicit name. Allowing an optional name makes the parser harder to write because we now need to know if "with" is a name or a mixin, or maybe we need to make "with" a keyword. We already added the ability to influence the generated suffix, so I think that's going far enough for people to migrate. Anyone else that uses completely bespoke naming patterns simply can't use inline input and output definitions.